frida/src/script.rs

```diff
@@ -5,46 +5,167 @@
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type")]
+#[serde(rename_all = "lowercase")]
+pub enum Message {
+    /// Message of type "send"
+    Send(MessageSend),
+
+    /// Message of type "log"
+    Log(MessageLog),
+
+    /// Message of type "error"
+    Error(MessageError),
+
+    /// Any other type of message.
+    Other(Value),
+}
+
+/// Send Message.
+#[derive(Deserialize, Debug)]
+pub struct MessageSend {
+    /// Payload of a Send Message.
+    pub payload: SendPayload,
+}
+
+/// Log Message.
+#[derive(Deserialize, Debug)]
+pub struct MessageLog {
+    /// Log Level.
+    pub level: MessageLogLevel,
+    /// Payload of a Message Log.
+    pub payload: String,
+}
+
+/// Error message.
+/// This message is sent when a JavaScript runtime error occurs, such as a misspelled word.
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct MessageError {
+    /// Error description.
+    pub description: String,
+
+    /// Stack trace string.
+    pub stack: String,
+
+    /// Script file name that failed.
+    pub file_name: String,
+
+    /// Line number with the error.
+    pub line_number: usize,
+
+    /// Column number with the error.
+    pub column_number: usize,
+}
+
+/// Represents a Message Log Level Types.
+/// Used by `MessageLog._level`
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "lowercase")]
+pub enum MessageLogLevel {
+    /// Indicates an informal message.
+    Info,
+    /// Represents a debugging message.
+    Debug,
+    /// Signifies a warning message.
+    Warning,
+    /// Represents an error message.
+    Error,
+}
+
+/// Represents a Message Log Level Types.
+/// Used by `MessageLog._level`
+#[derive(Deserialize, Debug)]
+pub struct SendPayload {
+    /// Send message type
+    pub r#type: String,
+    /// Send message ID
+    pub id: usize,
+    /// Send message result.
+    pub result: String,
+    /// Send message returns.
+    pub returns: Value,
+}
+
 unsafe extern "C" fn call_on_message<I: ScriptHandler>(
     _script_ptr: *mut _FridaScript,
     message: *const i8,
     _data: &frida_sys::_GBytes,
     user_data: *mut c_void,
 ) {
-    let handler: &mut I = &mut *(user_data as *mut I);
+    let c_msg = CStr::from_ptr(message as *const c_char)
+        .to_str()
+        .unwrap_or_default();
+
+    let formated_msg: Message = serde_json::from_str(c_msg).unwrap_or_else(|err| {
+        Message::Other(serde_json::json!({
+            "error": err.to_string(),
+            "data": c_msg
+        }))
+    });
 
-    handler.on_message(
-        CStr::from_ptr(message as *const c_char)
-            .to_str()
-            .unwrap_or_default(),
-    );
+    match formated_msg {
+        Message::Send(msg) => {
+            if msg.payload.r#type == "frida:rpc" {
+                let callback_handler: *mut CallbackHandler = user_data as _;
+                on_message(callback_handler.as_mut().unwrap(), Message::Send(msg));
+            }
+        }
+        Message::Log(msg) => {
```

Comment:
I'd collapse the last three arms of this match to a single arm, as they are identical.

Reply:
On it.